### PR TITLE
fix(mlx): restore vision LoRA training and adapt role-free projections

### PR DESCRIPTION
**Pull request by Lyxot**
Follow-up to #1150. Selecting and wrapping vision adapters does not guarantee that image features reach the training loss: processor fallbacks can discard images or their markers, and a legacy single-placeholder processor can leave hundreds of projected features without matching token positions. This preserves the role-based selection from #1150 and repairs those shared input paths.

### Changes

- Preserve native image markers, including numbering across conversation turns; normalize processor `images` output to `pixel_values`; exclude negative image placeholders from generated language labels while retaining explicit-label validation.
- Recover installed native mlx-vlm processors when upstream silently returns a bare tokenizer, with a scoped retry for the tokenizer regex-repair incompatibility. Preserve tokenizer trust scopes and refusal propagation.
- For legacy bare-tokenizer processors, measure the projected image-token count without retaining features, expand complete image spans before label/width planning, and recompute features inside the training gradient closure. Reject split spans and incompatible unpatched merges; composed native processors retain their own preprocessing.
- When a vision tower has no role-selected projections and both attention/MLP flags are enabled, adapt supported flat linears and plain ungrouped 1×1 convolution projections. Convolution adapters support training, reload, and fusion. Other convolution kernels stay frozen and empty eligible groups still raise.
- Refuse selected BitLinear targets before conversion with a diagnostic naming the packed-weight CustomKernel's missing input VJP. Downstream head-only adapters remain supported. Preserve upstream DoRA preflight and conversion behavior.

### Validation

After rebasing onto current `main`, **107 focused tests pass**: 86 selection/seed/last-layer tests, 7 DoRA tests, 12 processor trust/refusal tests, and 2 BitLinear gradient cases on Metal. CPU selection tests do not execute Metal-only BitLinear kernels, and optional upstream BitLinear variants are checked when present in the installed mlx-vlm version.

```sh
python -m pytest tests/test_mlx_lora_group_selection.py tests/test_mlx_get_peft_model_seed_ordering.py tests/test_mlx_finetune_last_n_layers.py -q
python -m pytest tests/test_mlx_training_e2e_metal.py -k bitlinear -q
python -m pytest tests/test_mlx_switch_lora.py -k dora -q
python -m pytest tests/test_mlx_save_export_regressions.py -k 'processor_loader_is_call_scoped or tokenizer_scope_forwards or processor_remote_refusal or swallowed_processor_refusal or image_processor_builder_forwards' -q
```

Before the rebase, tiny real checkpoints exercised `from_pretrained → get_peft_model → trainer` for two image steps, counting **only LoRA-B movement**:

| Family | Vision LoRA-B leaves moved |
|---|---:|
| aya_vision | 12 |
| glm5_next | 10 |
| phi3_v | 138 |
| step3p7 | 8 |
| mistral4 | 14 |
| llava_bunny | 12 |
| mllama | 24 |
| deepseekocr | 8 |
| deepseekocr_2 | 168 |
| gemma3n / gemma4_unified | 233 / 1 |
| kimi_k3 / kimi_vl | 8 / 8 |
| molmo / molmo2 | 20 / 26 |

All nine B4 runs had CE/CCE loss divergence 0.0. The persisted shared-merge regression fails when the image merge is disconnected. Additional mutation checks cover marker/pixel preservation, projection selection, convolution training/fusion, native recovery, and span validation. Native Bunny checks preserve all 729 image slots across two ragged rows, both padding directions, and plain/prompt-completion formats.

The image results use a 1024-token cap: Aya and Mistral4 already worked at that length because 512 truncated away their images. Mllama required a nonzero frozen gate in its synthetic checkpoint, and DeepSeekOCR required consistent synthetic tower/projector dimensions; neither correction changes production weights or gates. GLM5Next was validated with Transformers 5.16.0/tokenizers 0.23.1 plus the authoritative upstream processor classes; this PR does not install those dependencies. The full family matrix was not repeated after the rebase.

### CODE REVIEW

**Comment by Lyxot** [comment]
@codex review

**Comment by Lyxot** [comment]
@codex review

**Comment by danielhanchen** [comment]
Confirmed the image merge and processor fallback paths this touches, and that a roleless vision tower currently raises at _raise_group_empty in unsloth_zoo/mlx/loader.py rather than adapting, so the default vision flags will now train more parameters and cost memory and step time. Could you split the BitLinear refusal and the pointwise convolution adapters out from the image-merge fix, and confirm the widened default selection is intended?

**Comment by danielhanchen** [comment]
I rebased this onto current main myself and pushed the resolution, so the PR is mergeable again, then ran it on real Apple Silicon and against a main control on three OSes. Summary: the defect is real, it is narrower than the title suggests, and I could not find anything it breaks.

## What the resolution changed

Four of the five conflicts with #1184 were add/add adjacency and keep both sides (`_vlm_token_messages` next to `_mark_vlm_image_parts`, the legacy-image block next to `_ensure_vlm_pad_token`, the `renderer` probe, and `legacy_image_inputs` short-circuiting ahead of the tokenizer/images/audio locals, which are mutually exclusive with it).

The fifth was not inside a conflict marker and is worth calling out. `root_modules` reads `model.named_modules()`, while `root.update_modules()` below it uses the root #1184 resolves by walking `.model` until it finds `named_modules`/`update_modules`. Git merged that line cleanly because #1184 never touched it, so on a VLM the names collected from `model` would not have resolved against `root` and #1184 would have been silently reverted for the root pass. `root_modules` now walks `root`.

## Is the bug real

Yes, and it is one key. Running your test files against unmodified main with the real mlx-cpu runtime:

| Test | main | this PR |
|---|---|---|
| `test_vision_lora_b_moves_through_the_processed_image_merge[pixel_values]` | passes | passes |
| `test_vision_lora_b_moves_through_the_processed_image_merge[images]` | **fails** | passes |

So vision LoRA-B was reaching the loss all along when the processor returned `pixel_values`, and never when it returned `images`. That is the `_to_mx_vlm_batch` alias, and it is a genuine fix. 17 of your tests fail on main and all pass with the PR, and those three files have zero baseline failures on main, so all 17 are signal.

I would retitle around that. "Restore vision LoRA training" reads as though vision training was broken outright, and the evidence says it was broken for one processor output shape.

## Does it break anything

Base versus head, identical lanes, real runtimes:

| Lane | main | this PR | verdict |
|---|---|---|---|
| Apple Silicon `macos-15`, repo's own `mlx-ci.yml`, 21 files, per-file processes | not previously run on this PR | **green, ~1000 tests** | pass |
| Linux x86_64, real mlx-cpu | 273 passed | 292 passed (+19 new) | pass |
| Windows Server 2025 | 1 failed, 217 passed | 1 failed, 217 passed | identical |
| macOS arm64, PR's four files | 1 failed | 1 failed | identical |
| main's 8 neighbouring MLX suites on Linux | 491 passed, 6 skipped | 491 passed, 6 skipped | identical |

The two red cells are the same single test on both branches, so neither is yours:

- Windows: `test_trusted_dir_handles_a_root_trusted_path` fails on main too. `env_value=os.sep` is `\` on Windows and the root-path handling does not hold there.
- macOS: `test_image_processor_builder_forwards_trust[False]` fails on main too.

Both files are only ever run on Linux today, which is why neither has been noticed. I will open those separately.

## The default-selection question I raised earlier

Answered, and the answer is narrower than I assumed. Exact inventories:

| Case | main | this PR |
|---|---|---|
| Roleless vision tower, default flags | raises at `_raise_group_empty` | trains `blocks.0.patch_dense`, 4480 params |
| Roleless tower, `finetune_attention_modules=False` | raises | raises |
| Roleless tower, `finetune_mlp_modules=False` | raises | raises |
| Roleless tower, explicit `target_modules` | `['blocks.0.patch_dense']` | `['blocks.0.patch_dense']` |
| Role-bearing tower, default flags | 4 modules, 5376 params | 4 modules, 5376 params |
| Plain text model, defaults and `all-linear` | 8 modules, 4096 params | 8 modules, 4096 params |

No configuration that works today selects differently. Only configurations that previously raised now train, and only when both group flags are on. That is enough for me, so ignore my earlier request to split the selection change out.

## BitLinear is not a compatibility break

I had this down as the biggest risk and it is not one. On main a BitNet model already raised, with `No MLX LoRA target modules were found ... use target_modules='all-linear'`, and `all-linear` raised the same thing, so the hint sent people nowhere. The PR replaces it with an accurate diagnostic. `target_modules=['lm_head']` adapts `lm_head` with 768 params identically on both branches, and the refusal leaves trainable parameters and attached adapters untouched either way.

## Checkpoint compatibility

Text adapters round trip both directions with identical tensor names, old-saved to new-loaded and new-saved to old-loaded. Adapters containing the new roleless or pointwise-conv targets can only be produced by this code, and an older install fails loudly when rebuilding the model rather than loading a partial adapter, so there is no silent-corruption path.

## Two small things

`_vlm_token_messages` catches only `ValueError` from `get_message_json`. In mlx-vlm 0.6.4 an unknown model type raises exactly that and `None`/`""` are already short-circuited by the `if not model_type` guard, so this is correct today. It is one upstream refactor away from becoming a `KeyError` that escapes into every VLM render path, and the advisory `latest` leg is where that would first show up.

I have also added the `mlx` label, so the Apple Silicon lane runs on this PR on every push from now on rather than never.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex security review

File: unsloth_zoo/mlx/utils.py [hunk 1]
```diff
@@ -4746,6 +4755,49 @@ def _processor_accepts_assistant_list_content(processor):
...
+            continue
+        text = "".join(str(part.get("text", "")) for part in parts
+                       if part.get("type") == "text")
+        if hasattr(processor, "_unsloth_legacy_image_spec"):
+            rendered.append({**message, "content": "".join(
+                "<image>\n" if part["type"] == "image" else str(part.get("text", "")) for part in parts)})
+            continue
+        try:
+            native = get_message_json(
+                model_type, text, role=message.get("role", "user"), num_images=count,
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/utils.py:4785]: Checked this against the code and against mlx-vlm's own API, and it does not hold.

The rewrite you describe only happens when `get_message_json` returns **string** content. The guard is `isinstance(content, str)` at utils.py:4798, and for every interleaving-capable family `get_message_json` returns a **list**, so the message is passed through untouched. Enumerating the builder over all model types it knows, on mlx-vlm 0.6.4:

- list content, left unchanged: `qwen2_vl`, `qwen2_5_vl`, `qwen3_vl`, `qwen3_vl_moe`, `llava`, `pixtral`, `idefics2`, `idefics3`, `smolvlm`, `gemma4`, `gemma4_unified`, `gemma3n`, `mistral3`, `llama4`, `molmo2`, `glm4v`, `internvl_chat`, `kimi_vl`, and 23 more (41 total)
- string content, rewritten: 17 families, `phi3_v`, `gemma3`, `deepseekocr`, `deepseek_vl_v2`, `jina_vlm`, `minicpmv4_6`, `step3p7`, `phi4mm` among them

Feeding your exact example, `First:` / image / `Second:` / image:

```
qwen2_vl     unchanged=True    (original list preserved)
qwen3_vl     unchanged=True    (original list preserved)
llava        unchanged=True    (original list preserved)
pixtral      unchanged=True    (original list preserved)
idefics3     unchanged=True    (original list preserved)
gemma4       unchanged=True    (original list preserved)
phi3_v       unchanged=False   <|image_1|><|image_2|>First:Second:
gemma3       unchanged=False   First:Second:<start_of_image><start_of_image>
deepseekocr  unchanged=False   <image>\n<image>\nFirst:Second:
```

For the 17 that are rewritten, the block form is the format those models expect, and the code reproduces the side the family actually uses: markers before the text for phi3_v and deepseekocr, after it for gemma3. Microsoft's own Phi-3-vision reference builds the prompt as `<|user|>\n<|image_1|>\n<|image_2|>\n{text}<|end|>`, all placeholders first, and the model card notes the checkpoint is designed for a single image per prompt in the first place.

The remedy is also not available. `get_message_json(model_type, text, role, num_images)` takes a count and no positions, so for exactly the families where this code fires there is no way to ask it for per-image placement. Interleaving is expressible only through the list-content families, and those are the ones the guard already skips.

**Comment by danielhanchen** [comment]
@codex review
